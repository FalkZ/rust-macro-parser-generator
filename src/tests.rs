macro_rules! tests {
    (@debug: $name:ident, $t:ident, $str:literal) => {
        #[test]
        fn $name() {
            let r = crate::m1n::renderer::render($str).expect("couldn't compile file");
            insta::assert_debug_snapshot!(&r.$t);
        }
    };
    (@display: $name:ident, $t:ident, $str:literal) => {
        #[test]
        fn $name() {
            let r = crate::m1n::renderer::render($str).expect("couldn't compile file");
            insta::assert_display_snapshot!(&r.$t);
        }
    };
    // `()` indicates that the macro takes no argument.
    ($($name:ident: $str:literal),*) => {
        mod tokens {
            $(
                tests!(@debug: $name, tokens, $str);
            )*
        }
        mod statements {
            $(
                tests!(@debug: $name, statements, $str);
            )*
        }
        mod typescript {
            $(
                tests!(@display: $name, typescript, $str);
            )*
        }
    };
}

tests!(
    enums: "./examples/m1n/Enum.m1n",
    class: "./examples/m1n/Class.m1n",
    file: "./examples/m1n/file.m1n",
    math: "./examples/m1n/std/math.m1n",
    matcher: "./examples/m1n/match.m1n",
    assignment: "./examples/m1n/assignment.m1n",
    operator: "./examples/m1n/operator.m1n"
);
