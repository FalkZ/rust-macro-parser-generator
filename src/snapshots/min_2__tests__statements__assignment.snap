---
source: src/tests.rs
expression: "&r.statements"
---
[
    statements {
        position: Position {
            column: 0,
            line: 1,
        },
        statement: imports(
            imports {
                position: Position {
                    column: 0,
                    line: 1,
                },
                imports: [
                    import {
                        position: Position {
                            column: 4,
                            line: 2,
                        },
                        path: RAWIDENT(
                            Token {
                                raw: "`@std/util`",
                                position: Position {
                                    column: 4,
                                    line: 2,
                                },
                            },
                        ),
                        import_items: Some(
                            import_items {
                                position: Position {
                                    column: 15,
                                    line: 2,
                                },
                                items: [
                                    import_item {
                                        position: Position {
                                            column: 17,
                                            line: 2,
                                        },
                                        item: IDENT(
                                            Token {
                                                raw: "pipe",
                                                position: Position {
                                                    column: 17,
                                                    line: 2,
                                                },
                                            },
                                        ),
                                    },
                                ],
                            },
                        ),
                    },
                    import {
                        position: Position {
                            column: 4,
                            line: 3,
                        },
                        path: RAWIDENT(
                            Token {
                                raw: "`@std/math`",
                                position: Position {
                                    column: 4,
                                    line: 3,
                                },
                            },
                        ),
                        import_items: None,
                    },
                ],
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 7,
        },
        statement: variable(
            variable {
                position: Position {
                    column: 0,
                    line: 7,
                },
                modifiers: [
                    modifiers {
                        position: Position {
                            column: 0,
                            line: 7,
                        },
                        modifier: MUT(
                            Token {
                                raw: "mut",
                                position: Position {
                                    column: 0,
                                    line: 7,
                                },
                            },
                        ),
                    },
                ],
                name: IDENT(
                    Token {
                        raw: "val",
                        position: Position {
                            column: 4,
                            line: 7,
                        },
                    },
                ),
                body: body {
                    position: Position {
                        column: 10,
                        line: 7,
                    },
                    value: primitive_value(
                        number(
                            number {
                                position: Position {
                                    column: 10,
                                    line: 7,
                                },
                                negative: None,
                                whole: NUMBER(
                                    Token {
                                        raw: "0",
                                        position: Position {
                                            column: 10,
                                            line: 7,
                                        },
                                    },
                                ),
                            },
                        ),
                    ),
                    expressions: [],
                },
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 9,
        },
        statement: function(
            function {
                position: Position {
                    column: 0,
                    line: 9,
                },
                modifiers: [],
                name: IDENT(
                    Token {
                        raw: "setVal",
                        position: Position {
                            column: 0,
                            line: 9,
                        },
                    },
                ),
                arguments: Some(
                    arguments {
                        position: Position {
                            column: 7,
                            line: 9,
                        },
                        arguments: [],
                        last: IDENT(
                            Token {
                                raw: "v",
                                position: Position {
                                    column: 7,
                                    line: 9,
                                },
                            },
                        ),
                    },
                ),
                body: body {
                    position: Position {
                        column: 4,
                        line: 10,
                    },
                    value: primitive_value(
                        IDENT(
                            Token {
                                raw: "v",
                                position: Position {
                                    column: 4,
                                    line: 10,
                                },
                            },
                        ),
                    ),
                    expressions: [
                        expressions {
                            position: Position {
                                column: 6,
                                line: 10,
                            },
                            expression: assingment_operation(
                                assingment_operation {
                                    position: Position {
                                        column: 6,
                                        line: 10,
                                    },
                                    identifier: IDENT(
                                        Token {
                                            raw: "val",
                                            position: Position {
                                                column: 9,
                                                line: 10,
                                            },
                                        },
                                    ),
                                },
                            ),
                        },
                    ],
                },
            },
        ),
    },
]
